blockstore/splitstore/debug.go

```diff
@@ -0,0 +1,360 @@
+			break
+		}
+	}
+}
+
+func (d *debugLog) Flush() {
+	if d == nil {
+		return
+	}
+
+	// rotate non-empty logs
+	d.rotateReadLog()
+	d.rotateWriteLog()
+	d.rotateDeleteLog()
+	d.rotateStackLog()
+}
+
+func (d *debugLog) rotateReadLog() {
+	d.readMx.Lock()
+	defer d.readMx.Unlock()
+
+	if d.readCnt == 0 {
+		return
+	}
+
+	err := d.rotate(d.readLog, d.readPath)
+	if err != nil {
+		log.Warnf("error rotating read log: %s", err)
+		return
+	}
+
+	d.readLog, err = os.OpenFile(d.readPath, os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		log.Warnf("error opening log file: %s", err)
+		return
+	}
+
+	d.readCnt = 0
+}
+
+func (d *debugLog) rotateWriteLog() {
+	d.writeMx.Lock()
+	defer d.writeMx.Unlock()
+
+	if d.writeCnt == 0 {
+		return
+	}
+
+	err := d.rotate(d.writeLog, d.writePath)
+	if err != nil {
+		log.Warnf("error rotating write log: %s", err)
+		return
+	}
+
+	d.writeLog, err = os.OpenFile(d.writePath, os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		log.Warnf("error opening write log file: %s", err)
+		return
+	}
+
+	d.writeCnt = 0
+}
+
+func (d *debugLog) rotateDeleteLog() {
+	d.deleteMx.Lock()
+	defer d.deleteMx.Unlock()
+
+	if d.deleteCnt == 0 {
+		return
+	}
+
+	err := d.rotate(d.deleteLog, d.deletePath)
+	if err != nil {
+		log.Warnf("error rotating delete log: %s", err)
+		return
+	}
+
+	d.deleteLog, err = os.OpenFile(d.deletePath, os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		log.Warnf("error opening delete log file: %s", err)
+		return
+	}
+
+	d.deleteCnt = 0
+}
+
+func (d *debugLog) rotateStackLog() {
+	d.stackMx.Lock()
+	defer d.stackMx.Unlock()
+
+	if d.stackCnt == 0 {
+		return
+	}
+
+	err := d.rotate(d.stackLog, d.stackPath)
+	if err != nil {
+		log.Warnf("error rotating stack log: %s", err)
+		return
+	}
+
+	d.stackLog, err = os.OpenFile(d.stackPath, os.O_WRONLY|os.O_CREATE, 0644)
+	if err != nil {
+		log.Warnf("error opening stack log file: %s", err)
+		return
+	}
+
+	d.stackCnt = 0
+}
+
+func (d *debugLog) rotate(f *os.File, path string) error {
+	err := f.Close()
+	if err != nil {
+		return xerrors.Errorf("error closing file: %w", err)
+	}
+
+	arxivPath := fmt.Sprintf("%s-%d", path, time.Now().Unix())
+	err = os.Rename(path, arxivPath)
+	if err != nil {
+		return xerrors.Errorf("error moving file: %w", err)
+	}
+
+	go func() {
+		cmd := exec.Command("gzip", arxivPath)
+		err := cmd.Run()
+		if err != nil {
+			log.Warnf("error compressing log: %s", err)
+		}
+	}()
+
+	return nil
+}
+
+func (d *debugLog) Close() error {
+	if d == nil {
+		return nil
+	}
+
+	d.readMx.Lock()
+	err1 := d.readLog.Close()
+	d.readMx.Unlock()
+
+	d.writeMx.Lock()
+	err2 := d.writeLog.Close()
+	d.writeMx.Unlock()
+
+	d.deleteMx.Lock()
+	err3 := d.deleteLog.Close()
+	d.deleteMx.Unlock()
+
+	d.stackMx.Lock()
+	err4 := d.stackLog.Close()
+	d.stackMx.Unlock()
+
+	return multierr.Combine(err1, err2, err3, err4)
+}
+
+func (d *debugLog) getStack() string {
+	sk := d.getNormalizedStackTrace()
+	hash := sha256.Sum256([]byte(sk))
```

Comment:
Do we _really_ need a sha256 hash here? Just a _tad_ slow. Just a bit.

(probably not critical)

Reply:
No, even md5 will do -- but in the grand scheme of things we are taking a stack trace, which is a lot slower than the hash!

Reply:
Yeah, probably.